gsr_booking/booking_logic.py

```diff
@@ -0,0 +1,255 @@
+                    failed_members.append(member)
+
+            # if unbooked slots still remain and not all booking requests succeeded, loop through each member again
+            # but get reservation credits first to see how much we can book
+
+            for member in failed_members:
+                if next_end - next_start < datetime.timedelta(hours=0.1):
+                    print("BOOKED EVERYTHING ALREADY")
+                    break
+                if member["user__email"] is "":
+                    print("empty email address")
+                    continue
+
+                # calculate number of credits already used via getReservations
+                (success, used_credit_hours) = self.get_used_booking_credit_for_user(
+                    lid, member["user__email"]
+                )
+                remaining_credit_hours = self.MAX_SLOT_HRS - used_credit_hours
+                rounded_remaining_credit_hours = math.floor(2 * remaining_credit_hours) / 2
+                if success and remaining_credit_hours >= self.MIN_SLOT_HRS:
+                    next_end = min(
+                        END_DATE,
+                        next_start + datetime.timedelta(hours=rounded_remaining_credit_hours),
+                    )
+                    (success, error) = self.book_room_for_user(
+                        room,
+                        lid,
+                        next_start.strftime(self.DATE_FORMAT_STR),
+                        next_end.strftime(self.DATE_FORMAT_STR),
+                        member["user__email"],
+                    )
+                    if success:
+                        key = f"{lid}_{room}"
+                        if not key in bookings:
+                            bookings[key] = []
+                        new_bookings = self.split_booking(
+                            next_start, next_end, member["username"], True
+                        )
+                        bookings[key].extend(new_bookings)
+
+                        next_start = next_end
+                        next_end = min(
+                            END_DATE, next_start + datetime.timedelta(hours=MAX_SLOT_HRS)
+                        )
+                    elif self.is_fatal_error(error):
+                        result_json = self.construct_bookings_json_obj(
+                            bookings, lid, room, END_DATE, next_start, next_end, error
+                        )
+                        return result_json
+            result_json = self.construct_bookings_json_obj(
+                bookings, lid, room, END_DATE, next_start, next_end, error
+            )
+            return result_json
+
+    def is_fatal_error(self, error):
+        # consider all errors fatal unless its a daily limit error
+        if error is not None:
+            if "daily limit".lower() not in error.lower():
+                return False
+            return True
+        return False
+
+    def construct_bookings_json_obj(
```

Comment:
Would be cool to write a docstring here that shows what a booking json object looks like, just so devs don't have to construct it in their head.